tools/cli/src/mentat_cli/input.rs

```diff
@@ -71,6 +78,44 @@ impl InputReader {
         self.reader.is_some()
     }
 
+    pub fn read_file(&mut self, filename: &String) -> Result<Vec<InputResult>, cli::Error> {
+        let path = Path::new(&filename);
+        let display = path.display();
+
+        let f = match File::open(&path) {
+            Err(err) => bail!(cli::ErrorKind::FileError(display.to_string(), err.to_string())),
+            Ok(file) => file,
+        };
+
+        let mut buffer = String::new();
+
+        let file = BufReader::new(&f);
+        let mut cmd_err: Option<edn::ParseError> = None;
+
+        let mut results: Vec<InputResult> = Vec::new();
+        for line in file.lines() {
+            let l = line.unwrap();
+            println!("{}", l);
+            buffer.push_str(&l);
+            let cmd = Command::Transact(buffer.to_string());
+            let (is_complete, err) = cmd.is_complete();
+            if is_complete {
+                results.push(InputResult::MetaCommand(cmd));
+                buffer.clear();
+                cmd_err = None;
+            } else {
+                cmd_err = err;
+            }
+
+        }
+
+        if let Some(err) = cmd_err {
+            bail!(err);
```

Comment:
`bail!` is not lexically scoped; so you can `bail!` just like you can `return`.  There's no need to save this potential error and re-throw it.